### PR TITLE
Aux/node locator

### PR DESCRIPTION
**Pull request by sshaman1101**
This PR adds simple test service that allows SONM nodes to discover each other by their Etherum addr.

### CODE REVIEW

File: insonmnia/hub/server.go [hunk 1]
```diff
@@ -601,3 +611,44 @@ func (h *Hub) deleteTaskByID(taskID string) {
 	defer h.tasksmu.Unlock()
 	delete(h.tasks, taskID)
 }
+
+func (h *Hub) startLocatorAnnouncer() {
+	tk := time.NewTicker(300 * time.Second)
```
Comment (3Hren) [inline insonmnia/hub/server.go]: Make configurable.

File: insonmnia/hub/server.go [hunk 2]
```diff
@@ -601,3 +611,44 @@ func (h *Hub) deleteTaskByID(taskID string) {
...
+	for {
+		select {
+		case <-tk.C:
+			h.announceAddress(h.ctx)
+		}
+	}
+}
+
+func (h *Hub) announceAddress(ctx context.Context) {
+	conn, err := grpc.Dial(h.locatorEndpoint,
```
Comment (3Hren) [inline insonmnia/hub/server.go]: No reason to perform connection for each attempt.

**Comment by 3Hren** [comment]
Any serious usages of this is expected? It's unefficient, fault-untolerant and only C from CAS is implemented, and that's only because of single server.

File: insonmnia/hub/config_test.go [hunk 3]
```diff
@@ -112,3 +121,47 @@ monitoring:
 
 	assert.Equal(t, 1, conf.Logging.Level)
 }
+
+func TestLoadConfigWithoutLocator(t *testing.T) {
+	err := createTestConfigFile(`
+ethereum:
+  private_key: "1000000000000000000000000000000000000000000000000000000000000000"
```
Comment (noxiouz) [inline insonmnia/hub/config_test.go]: Better set a path to a private key file

File: insonmnia/hub/server.go [hunk 4]
```diff
@@ -861,3 +880,48 @@ func (h *Hub) deleteTask(taskID string) error {
...
+func (h *Hub) startLocatorAnnouncer() {
+	tk := time.NewTicker(h.locatorPeriod)
+	defer tk.Stop()
+
+	h.announceAddress(h.ctx)
+
+	for {
+		select {
+		case <-tk.C:
+			h.announceAddress(h.ctx)
```
Comment (noxiouz) [inline insonmnia/hub/server.go]: Case context.Done. otherwise it loops forever. Can't be stopped

**Comment by noxiouz** [comment]
How do u get those "merge brach"? Force push...

File: insonmnia/hub/server.go (hunk 4 above)
Comment (antmat) [inline insonmnia/hub/server.go]: Do we really need to do it in cycle? Maybe we can track connection from hub to locator

File: insonmnia/locator/server.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	mx sync.Mutex
+	db map[string]*node
+}
+
+func (l *Locator) putAnnounce(n *node) {
+	l.mx.Lock()
+	defer l.mx.Unlock()
+
+	n.ts = time.Now()
+	l.db[n.ethAddr] = n
```
Comment (antmat) [inline insonmnia/locator/server.go]: What if several hubs have same ip?

File: insonmnia/hub/server.go [hunk 6]
```diff
@@ -781,8 +794,14 @@ func (h *Hub) Serve() error {
 			go h.handleInterconnect(h.ctx, conn)
 		}
 	}()
-	h.wg.Wait()
 
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		h.startLocatorAnnouncer()
```
Comment (antmat) [inline insonmnia/hub/server.go]: this one should be only on leader

**Comment by antmat** [comment]
still not merging

File: insonmnia/hub/server.go [hunk 7]
```diff
@@ -939,8 +946,23 @@ func (h *Hub) Serve() error {
 			go h.handleInterconnect(h.ctx, conn)
 		}
 	}()
-	h.wg.Wait()
 
+	// init locator connection and announce
+	// address only on Leader
+	if h.isLeader {
```
Comment (antmat) [inline insonmnia/hub/server.go]: and what happens when leadership is lost?

**Comment by antmat** [comment]
WTF?

**Pull request merged**